src/messages/MOSDPing.h

```diff
@@ -81,40 +79,61 @@ class MOSDPing : public Message {
     bufferlist::iterator p = payload.begin();
     ::decode(fsid, p);
     ::decode(map_epoch, p);
-    ::decode(peer_as_of_epoch, p);
-    ::decode(op, p);
-    ::decode(peer_stat, p);
+    if (header.version < 3) {
+      osd_peer_stat_t peer_stat;
+      epoch_t peer_as_of_epoch;
+      ::decode(peer_as_of_epoch, p);
+      ::decode(op, p);
+      ::decode(peer_stat, p);
+    } else {
+      ::decode(op, p);
+    }
     ::decode(stamp, p);
     if (header.version >= 3) {
-      bufferlist size_bl;
       int payload_mid_length = p.get_off();
-      ::decode(size_bl, p);
-      min_message_size = size_bl.length() + payload_mid_length;
+      uint32_t size;
+      ::decode(size, p);
+      p.advance(size);
+      min_message_size = size + payload_mid_length;
     }
   }
   void encode_payload(uint64_t features) override {
     ::encode(fsid, payload);
     ::encode(map_epoch, payload);
-    ::encode(peer_as_of_epoch, payload);
-    ::encode(op, payload);
-    ::encode(peer_stat, payload);
-    ::encode(stamp, payload);
-
-    bufferlist pad;
-    size_t s = MAX(min_message_size - payload.length(), 0);
-    // this should be big enough for normal min_message padding sizes. since
-    // we are targetting jumbo ethernet frames around 9000 bytes, 16k should
-    // be more than sufficient!  the compiler will statically zero this so
-    // that at runtime we are only adding a bufferptr reference to it.
-    static char zeros[16384] = {};
-    while (s > sizeof(zeros)) {
-      pad.append(buffer::create_static(sizeof(zeros), zeros));
-      s -= sizeof(zeros);
-    }
-    if (s) {
-      pad.append(buffer::create_static(s, zeros));
+    
+    // with luminous, we drop peer_as_of_epoch and peer_stat, and we add extra 
+    // payload to artificially inflate the ping packet size.
+    if (!HAVE_FEATURE(features, SERVER_LUMINOUS)) { 
+      epoch_t dummy_epoch = {};
+      osd_peer_stat_t dummy_stat = {};
+      header.version = 2;
+      ::encode(dummy_epoch, payload);
+      ::encode(op, payload);   
+      ::encode(dummy_stat, payload);
+      ::encode(stamp, payload);
+    } else {
+      ::encode(op, payload);
+      ::encode(stamp, payload);
+      size_t s = 0;
+      if (min_message_size > payload.length())
+        s = min_message_size - payload.length();
+      if (s) {
+        // this should be big enough for normal min_message padding sizes. since
+        // we are targetting jumbo ethernet frames around 9000 bytes, 16k should
+        // be more than sufficient!  the compiler will statically zero this so
+        // that at runtime we are only adding a bufferptr reference to it.
+        static char zeros[16384] = {};
+        while (s > sizeof(zeros)) {
+          payload.append(buffer::create_static(sizeof(zeros), zeros));
+          s -= sizeof(zeros);
+        }
+        if (s) {
+          payload.append(buffer::create_static(s, zeros));
+        }
+      } else {
+        ::encode((uint32_t)s, payload);
```

Comment:
Good catch, fixing. It didn't crash on me probably because of protocol robustness...